### PR TITLE
[SPARK-43803] [SS] [CONNECT] Improve awaitTermination() to handle client disconnects

### PR DESCRIPTION
**Pull request by pengzhon-db**
### What changes were proposed in this pull request?

Streaming awaitTermination() is a long running API. Currently, it keeps running on server even if client disconnects. This change periodically checks if client has disconnected. If so, we can stop the operation and release resources.

We use gRPC Context.isCancelled() to determine if client has disconnected and response cannot be returned to the client. [Here is the reference of of isCancelled()](https://grpc.github.io/grpc/cpp/classgrpc_1_1_server_context.html#af2d0f087805b4b475d01b12d73508f09): 

> Return whether this RPC failed before the server could provide its status back to the client.

> This could be because of explicit API cancellation from the client-side or server-side, because of deadline exceeded, network connection reset, HTTP/2 parameter configuration (e.g., max message size, max connection age), etc. It does NOT include failure due to a non-OK status return from the server application's request handler, including [Status::CANCELLED](https://grpc.github.io/grpc/cpp/classgrpc_1_1_status.html#a9994ffe95a0495915d82481c2ec594ab).

> IsCancelled is always safe to call when using sync or callback API. When using async API, it is only safe to call IsCancelled after the AsyncNotifyWhenDone tag has been delivered. Thread-safe.

### Why are the changes needed?

The change improves handling of awaitTermination(). It avoids resource waste of server side.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing unit tests.
Manually tested on local with:
1. start spark connect 
2. create a streaming query
3. call query.awaitTermination()
4. exit() the client to disconnect it
5. check that an error (RPC context is cancelled when executing awaitTermination()) is logged on server. It proves that awaitTermination() is exited on the server side when client disconnects

```
>>> query = (
...  spark
...  .readStream
...  .format("rate")
...  .option("numPartitions", "1")
...  .load()
...  .writeStream
...  .format("memory")
...  .queryName("tableName_35")
...  .start()
... )

>>>
>>> query.awaitTermination()
...
>>> exit()
```

### CODE REVIEW

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 1]
```diff
@@ -2597,6 +2600,50 @@ class SparkConnectPlanner(val session: SparkSession) {
...
+      if (hasTimeout) {
+        timeoutLeftMs -= awaitTerminationIntervalMs
+        if (timeoutLeftMs <= 0) {
+          return false
+        }
+      }
+    }
+
+    // gRPC is cancelled
+    logError("RPC context is cancelled when executing awaitTermination()")
```
Comment (rangadi) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2643]: Should be warning at most. Info is fine too. Log the query id.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 2]
```diff
@@ -2597,6 +2600,50 @@ class SparkConnectPlanner(val session: SparkSession) {
...
+        awaitTerminationIntervalMs
+      }
+
+      val terminated = query.awaitTermination(awaitTimeMs)
+      if (terminated) {
+        return true
+      }
+
+      if (hasTimeout) {
+        timeoutLeftMs -= awaitTerminationIntervalMs
```
Comment (rangadi) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2635]: Subtract actual time elapsed.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 3]
```diff
@@ -2597,6 +2600,50 @@ class SparkConnectPlanner(val session: SparkSession) {
...
+    // How often to check if RPC is cancelled and call awaitTermination()
+    val awaitTerminationIntervalMs = 10000
+
+    val hasTimeout = timeoutMs.isDefined
+    var timeoutLeftMs = timeoutMs.getOrElse(Long.MaxValue)
+    require(timeoutLeftMs > 0, "Timeout has to be positive")
+
+    val grpcContext = Context.current
+    while (!grpcContext.isCancelled) {
+      val awaitTimeMs = if (hasTimeout) {
```
Comment (rangadi) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2623]: No need to check, timeout is already set to max value when it is not set.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 4]
```diff
@@ -2597,6 +2600,50 @@ class SparkConnectPlanner(val session: SparkSession) {
...
+   * @param timeoutMs optional. Timeout to wait for termination. If None, no timeout is set
+   * @return if the query has terminated
+   */
+  private def handleStreamingAwaitTermination(
+      query: StreamingQuery,
+      timeoutMs: Option[Long]): Boolean = {
+    // How often to check if RPC is cancelled and call awaitTermination()
+    val awaitTerminationIntervalMs = 10000
+
+    val hasTimeout = timeoutMs.isDefined
```
Comment (rangadi) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2617]: This is not needed.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 5]
```diff
@@ -2576,11 +2578,12 @@ class SparkConnectPlanner(val session: SparkSession) {
...
+          val terminated = handleStreamingAwaitTermination(query,
+            Some(command.getAwaitTermination.getTimeoutMs))
           respBuilder.getAwaitTerminationBuilder
             .setTerminated(terminated)
         } else {
-          query.awaitTermination()
+          handleStreamingAwaitTermination(query, None)
           respBuilder.getAwaitTerminationBuilder
             .setTerminated(true)
         }
```
Comment (rangadi) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:2589]: Can we simplify this? We don't need larger conditional.  Could just be:

```scala
val timeout = if (command.getAwaitTermination.hasTimeoutMs)
   Some(command.getAwaitTermination.getTimeoutMs) else None
val terminated = handleStreamingAwaitTermination(query, timeout)
  respBuilder.getAwaitTerminationBuilder
  .setTerminated(terminated)

**Review by rangadi — COMMENTED** [review]
Left a few comments. Can we add unit test for this?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 2 above)
Comment (pengzhon-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala]: done

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 3 above)
Comment (pengzhon-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala]: done

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 4 above)
Comment (pengzhon-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala]: removed

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 5 above)
Comment (pengzhon-db) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala]: done

**Comment by pengzhon-db** [comment]
> Left a few comments. Can we add unit test for this?

What kind of unit test are you referring to? We have existing Spark connect awaittermination unit test. Right now we don't have way to simulate client disconnect from python side. Do you think we should add some unit test for that?

**Comment by rangadi** [comment]
> What kind of unit test are you referring to? We have existing Spark connect awaittermination unit test. Right now we don't have way to simulate client disconnect from python side. Do you think we should add some unit test for that?

We could disconnect in the test (since we have access to client). Lets may be that could be part of PR for client side changes.

**Comment by pengzhon-db** [comment]
@HyukjinKwon Could u review this PR?

**Comment by HyukjinKwon** [comment]
Is this good to go @rangadi ?

**Review by rangadi — APPROVED** [review]
@HyukjinKwon LGTM. will check CI failure.

**Comment by rangadi** [comment]
@pengzhon-db please fix scala-format error.

**Comment by pengzhon-db** [comment]
@HyukjinKwon can you help merge this? Thanks

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**